

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=auto>



<head>
  <meta charset="UTF-8">
  <link rel="apple-touch-icon" sizes="76x76" href="/img/favicon.png">
  <link rel="icon" href="/img/favicon.png">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=5.0, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="description" content="">
  <meta name="author" content="John Doe">
  <meta name="keywords" content="">
  <meta name="description" content="Mark Renton 更新于: 2020-09-14  Entrez 是什么 玩转 entrez-direct内容简介 NCBI 的 Entrez 工具功能非常强大，既可以从 web 页面访问 NCBI 来查询，也可以利用 Entrez 提供的 Web Services 来实现诸多功能（编写第三方程序等）。此外 NCBI 还提供了 Entrez 的命令行工具 edirect1。edirect 全">
<meta property="og:type" content="article">
<meta property="og:title" content="entrez工具">
<meta property="og:url" content="http://example.com/2021/09/12/entrez/index.html">
<meta property="og:site_name" content="林国鹏">
<meta property="og:description" content="Mark Renton 更新于: 2020-09-14  Entrez 是什么 玩转 entrez-direct内容简介 NCBI 的 Entrez 工具功能非常强大，既可以从 web 页面访问 NCBI 来查询，也可以利用 Entrez 提供的 Web Services 来实现诸多功能（编写第三方程序等）。此外 NCBI 还提供了 Entrez 的命令行工具 edirect1。edirect 全">
<meta property="og:locale" content="zh_CN">
<meta property="article:published_time" content="2021-09-12T13:27:47.000Z">
<meta property="article:modified_time" content="2021-09-12T13:35:31.257Z">
<meta property="article:author" content="John Doe">
<meta name="twitter:card" content="summary_large_image">
  
  <title>entrez工具 - 林国鹏</title>

  <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4/dist/css/bootstrap.min.css" />


  <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/github-markdown-css@4/github-markdown.min.css" />
  <link  rel="stylesheet" href="/lib/hint/hint.min.css" />

  
    
    
      
      <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/highlight.js@10/styles/github-gist.min.css" />
    
  

  
    <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3/dist/jquery.fancybox.min.css" />
  


<!-- 主题依赖的图标库，不要自行修改 -->

<link rel="stylesheet" href="//at.alicdn.com/t/font_1749284_ba1fz6golrf.css">



<link rel="stylesheet" href="//at.alicdn.com/t/font_1736178_kmeydafke9r.css">


<link  rel="stylesheet" href="/css/main.css" />

<!-- 自定义样式保持在最底部 -->


  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    var CONFIG = {"hostname":"example.com","root":"/","version":"1.8.12","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"right","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"copy_btn":true,"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":false,"offset_factor":2},"web_analytics":{"enable":false,"baidu":null,"google":null,"gtag":null,"tencent":{"sid":null,"cid":null},"woyaola":null,"cnzz":null,"leancloud":{"app_id":null,"app_key":null,"server_url":null,"path":"window.location.pathname"}},"search_path":"/local-search.xml"};
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
<meta name="generator" content="Hexo 5.4.0"></head>


<body>
  <header style="height: 70vh;">
    <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand" href="/">
      <strong>林国鹏</strong>
    </a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/">
                <i class="iconfont icon-home-fill"></i>
                首页
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/">
                <i class="iconfont icon-archive-fill"></i>
                归档
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/">
                <i class="iconfont icon-category-fill"></i>
                分类
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/">
                <i class="iconfont icon-tags-fill"></i>
                标签
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/">
                <i class="iconfont icon-user-fill"></i>
                关于
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search">
              &nbsp;<i class="iconfont icon-search"></i>&nbsp;
            </a>
          </li>
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle">&nbsp;<i
                class="iconfont icon-dark" id="color-toggle-icon"></i>&nbsp;</a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

    <div class="banner" id="banner" parallax=true
         style="background: url('/img/default.png') no-repeat center center;
           background-size: cover;">
      <div class="full-bg-img">
        <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
          <div class="page-header text-center fade-in-up">
            <span class="h2" id="subtitle" title="entrez工具">
              
            </span>

            
              <div class="mt-3">
  
  
    <span class="post-meta">
      <i class="iconfont icon-date-fill" aria-hidden="true"></i>
      <time datetime="2021-09-12 21:27" pubdate>
        2021年9月12日 晚上
      </time>
    </span>
  
</div>

<div class="mt-1">
  
    <span class="post-meta mr-2">
      <i class="iconfont icon-chart"></i>
      11k 字
    </span>
  

  
    <span class="post-meta mr-2">
      <i class="iconfont icon-clock-fill"></i>
      
      
      35 分钟
    </span>
  

  
  
</div>

            
          </div>

          
        </div>
      </div>
    </div>
  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="d-none d-lg-block col-lg-2"></div>
    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div class="py-5" id="board">
          <article class="post-content mx-auto">
            <!-- SEO header -->
            <h1 style="display: none">entrez工具</h1>
            
            <div class="markdown-body">
              <p><a href="mailto:indexofire@gmail.com">Mark Renton</a> 更新于: 2020-09-14</p>
<hr>
<h2 id="Entrez-是什么"><a href="#Entrez-是什么" class="headerlink" title="Entrez 是什么"></a>Entrez 是什么</h2><hr>
<h2 id="玩转-entrez-direct"><a href="#玩转-entrez-direct" class="headerlink" title="玩转 entrez-direct"></a>玩转 entrez-direct</h2><p>内容简介</p>
<p><a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/">NCBI</a> 的 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/gquery/">Entrez</a> 工具功能非常强大，既可以从 web 页面访问 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/">NCBI</a> 来查询，也可以利用 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/gquery/">Entrez</a> 提供的 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/books/NBK25501/">Web Services</a> 来实现诸多功能（编写第三方程序等）。此外 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/">NCBI</a> 还提供了 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/gquery/">Entrez</a> 的命令行工具 edirect<a target="_blank" rel="noopener" href="https://indexofire.github.io/pathongs/book/C01_Common-Utility/03_entrez/#fn:1">1</a>。edirect 全名为 <strong>Entrez Direct</strong>，里面包含了一组各司其职的工具和脚本，通过 Linux 系统下的管道pipeline<a target="_blank" rel="noopener" href="https://indexofire.github.io/pathongs/book/C01_Common-Utility/03_entrez/#fn:4">2</a>功能联合使用这些命令行工具，在服务器上实现高效的进行 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/gquery/">Entrez</a> 的检索，抓取，过滤，排序等操作。</p>
<h3 id="1-安装"><a href="#1-安装" class="headerlink" title="1. 安装"></a>1. 安装</h3><h4 id="1-1-下载安装"><a href="#1-1-下载安装" class="headerlink" title="1.1 下载安装"></a>1.1 下载安装</h4><p>edirect 直接下载预编译包，添加到系统路径中即可。</p>
<figure class="highlight shell"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs shell"><span class="hljs-meta">#</span><span class="bash"> 下载 edirect 安装包</span><br><span class="hljs-meta">$</span><span class="bash"> wget ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/edirect.tar.gz</span><br><span class="hljs-meta"></span><br><span class="hljs-meta">#</span><span class="bash"> 将 edirect 安装在用户 app 目录下，可以根据需求自行选择安装路径</span><br><span class="hljs-meta">$</span><span class="bash"> tar zxvf edirect.tar.gz -C ~/apps</span><br><span class="hljs-meta"></span><br><span class="hljs-meta">#</span><span class="bash"> 运行安装脚本</span><br><span class="hljs-meta">$</span><span class="bash"> <span class="hljs-built_in">cd</span> ~/apps/edirect</span><br><span class="hljs-meta">$</span><span class="bash"> bash ./setup.sh</span><br><span class="hljs-meta"></span><br><span class="hljs-meta">#</span><span class="bash"> 添加软件安装目录到系统环境变量中</span><br><span class="hljs-meta">$</span><span class="bash"> <span class="hljs-built_in">echo</span> <span class="hljs-string">&quot;export PATH=&#x27;\$PATH:\$HOME/apps/edirect/&#x27;&quot;</span> &gt;&gt; ~/.bashrc</span><br><span class="hljs-meta">$</span><span class="bash"> <span class="hljs-built_in">source</span> ~/.bashrc</span><br></code></pre></div></td></tr></table></figure>

<p>安装说明</p>
<p>耐心等待一会，当看到下面字符时，表示安装成功。</p>
<figure class="highlight shell"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs shell">ENTREZ DIRECT HAS BEEN SUCCESSFULLY INSTALLED AND CONFIGURED<br></code></pre></div></td></tr></table></figure>

<p> 也可以不运行 <code>setup.sh</code> 直接使用 edirect 的一些工具，不过如果要使用全部功能，有些perl模块需要安装，运行 <code>setup.sh</code> 来自动完成 CPAN 的 modual 安装以及 go 版本的 xtract 的编译（不运行安装脚本会调用 perl 版本的 xtract，数据量大时速度略慢）。 </p>
<p>提示</p>
<p>对于部分操作系统如 <strong>ubuntu</strong> 等，<strong>uname</strong> 位于 /bin，而 xtract 接口调用脚本使用 /usr/bin/uname 来判断使用那个版本的 xtract，如果要在这些操作系统下正常使用 xtract，需要让程序在正确路径可识别，这里通过添加系统 link 到 /usr/bin/unname。</p>
<figure class="highlight shell"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs shell"><span class="hljs-meta">#</span><span class="bash"> 查看 uname 路径。</span><br><span class="hljs-meta">$</span><span class="bash"> <span class="hljs-built_in">which</span> uname</span><br><span class="hljs-meta"></span><br><span class="hljs-meta">#</span><span class="bash"> 如果是位于 /bin/ 下，将其链接到 /usr/bin/ 下即可。</span><br><span class="hljs-meta">$</span><span class="bash"> sudo ln -s /bin/uname /usr/bin/uname</span><br></code></pre></div></td></tr></table></figure>

<h4 id="1-2-conda-中安装"><a href="#1-2-conda-中安装" class="headerlink" title="1.2 conda 中安装"></a>1.2 conda 中安装</h4><p>如果系统通过 conda 进行软件管理，可以通过如下方式安装。</p>
<figure class="highlight cmake"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs cmake">$ conda <span class="hljs-keyword">install</span> entrez-direct<br></code></pre></div></td></tr></table></figure>

<p>不过 conda 安装的还是 perl 版本的 xtract，而不是 go 版本的，需要到 conda/pkgs 包根目录中的 edirect 相应目录，用 <code>setup.sh</code> 自行编译后生成 xtract.Linux(Linux 操作系统)，将其复制到 conda 的 bin 路径即可。go 版本的 xtract 可以用 -help 参数查看帮助。</p>
<h3 id="2-基本用法"><a href="#2-基本用法" class="headerlink" title="2. 基本用法"></a>2. 基本用法</h3><p>表. 最常用的工具参见下表</p>
<table>
<thead>
<tr>
<th align="left">工具名称</th>
<th align="left">工具用途</th>
<th align="left">常用参数</th>
</tr>
</thead>
<tbody><tr>
<td align="left"><strong>esearch</strong></td>
<td align="left">搜索命令，将所要检索的内容提交到 Entrez 中，返回相应的结果记录</td>
<td align="left">-db、-query</td>
</tr>
<tr>
<td align="left"><strong>efetch</strong></td>
<td align="left">下载 NCBI 数据库中的记录和报告并以相应格式打印输出</td>
<td align="left">-db、-id、-format、-mode</td>
</tr>
<tr>
<td align="left"><strong>einfo</strong></td>
<td align="left">获取目标结果在数据库中的信息</td>
<td align="left">-db、-dbs、-fields、-links</td>
</tr>
<tr>
<td align="left"><strong>elink</strong></td>
<td align="left">对目标结果在其他数据库中比配结果</td>
<td align="left">-db、-id、-related、-target、-name</td>
</tr>
<tr>
<td align="left"><strong>epost</strong></td>
<td align="left">上传 UIDs 或者 序列登记号</td>
<td align="left">-db、-id、-format、-input、-label</td>
</tr>
<tr>
<td align="left"><strong>efilter</strong></td>
<td align="left">对之前的检索结果进行过滤或限制</td>
<td align="left">-query、-sort、-field</td>
</tr>
<tr>
<td align="left"><strong>xtract</strong></td>
<td align="left">将esearch获得的 XML 格式结果转换成表格格式</td>
<td align="left">-pattern、-if、-block、-element、-sep、-filter</td>
</tr>
<tr>
<td align="left"><strong>esummary</strong></td>
<td align="left">获得 XML 格式的建立</td>
<td align="left">-db、-id、-format、-mode</td>
</tr>
<tr>
<td align="left"><strong>ecitmatch</strong></td>
<td align="left">统计引用数据</td>
<td align="left">-journal、-year、-volume、-page、-author</td>
</tr>
</tbody></table>
<h4 id="2-1-esearch"><a href="#2-1-esearch" class="headerlink" title="2.1 esearch"></a>2.1 esearch</h4><p><strong>esearch</strong> 负责检索并返回 xml 格式的命中情况。使用 esearch 命令，相当于平时你通过浏览器访问 NCBI，在 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/gquery/">GQuery</a> 的 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/nuccore/">nucletide</a> 数据库搜索见面搜素关键字 <code>salmonella</code>，返回命中结果。</p>
<figure class="highlight stata"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs stata">$ esearch -<span class="hljs-keyword">db</span> nuccore -<span class="hljs-keyword">query</span> <span class="hljs-string">&quot;salmonella&quot;</span><br></code></pre></div></td></tr></table></figure>

<p>终端里会返回命中结果：</p>
<figure class="highlight apache"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs apache"><span class="hljs-section">&lt;ENTREZ_DIRECT&gt;</span><br>  <span class="hljs-section">&lt;Db&gt;</span><span class="hljs-attribute">nuccore</span>&lt;/Db&gt;<br>  <span class="hljs-section">&lt;WebEnv&gt;</span><span class="hljs-attribute">NCID_1_153624467_130</span>.<span class="hljs-number">14</span>.<span class="hljs-number">18</span>.<span class="hljs-number">34</span>_<span class="hljs-number">9001</span>_<span class="hljs-number">1490828308</span>_<span class="hljs-number">1411373464</span>_<span class="hljs-number">0</span>MetA<span class="hljs-number">0</span>_S_MegaStore_F_<span class="hljs-number">1</span>&lt;/WebEnv&gt;<br>  <span class="hljs-section">&lt;QueryKey&gt;</span><span class="hljs-attribute">1</span>&lt;/QueryKey&gt;<br>  <span class="hljs-section">&lt;Count&gt;</span><span class="hljs-attribute">2093561</span>&lt;/Count&gt;<br>  <span class="hljs-section">&lt;Step&gt;</span><span class="hljs-attribute">1</span>&lt;/Step&gt;<br><span class="hljs-section">&lt;/ENTREZ_DIRECT&gt;</span><br></code></pre></div></td></tr></table></figure>

<p><strong>esearch</strong> 需要设置的参数有 <strong>-db</strong>（所要搜索的目标数据库）和 <strong>-query</strong>（所要搜索的关键词）。</p>
<p>结果默认返回 xml 格式的数据（参见 <code>...</code>），其中 中的数值表示检索命中的记录数，这与 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/">NCBI</a> 的 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/gquery/">Entrez</a> 网页界面里进行检索操作的结果是一致的。</p>
<p><strong>esearch</strong> 还有一个参数 -sort 可以根据数据库类型返回不同的排序方式。默认的排序方式是按照时间逆序排列的。想要得到准确的查询结果，优化 query 很重，可以参见 <a target="_blank" rel="noopener" href="https://indexofire.github.io/pathongs/book/C01_Common-Utility/03_entrez/#3-filtering-queries">Filter</a> 相关内容。</p>
<h4 id="2-2-efetch"><a href="#2-2-efetch" class="headerlink" title="2.2 efetch"></a>2.2 efetch</h4><p><strong>efetch</strong> 可以用来下载各个数据库的各种格式数据。你可以通过2种方式下载：</p>
<ol>
<li>用 esearch 获得命中记录后再下载。下面这个命令，就会下载所有结果并保存到 result.fasta 文件中。</li>
</ol>
<figure class="highlight stata"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs stata">$ esearch -<span class="hljs-keyword">db</span> nuccore -<span class="hljs-keyword">query</span> <span class="hljs-string">&quot;salmonella&quot;</span> -days 10 | \<br>&gt; efetch -<span class="hljs-keyword">format</span> fasta -<span class="hljs-keyword">db</span> nuccore &gt; result.fasta<br></code></pre></div></td></tr></table></figure>

<p>Warning</p>
<p>一定要注意搜索命中值的数量，关键词命中太多，数据量会太大，全部下载可能直接进入假死状态。应该通过设置参数来限定搜索结果。当不确定时，先用esearch工具看返回的命中值有多少再决定是否进一步限定。</p>
<ol>
<li>直接指定 id 下载</li>
</ol>
<figure class="highlight python"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs python">$ efetch -db pubmed -<span class="hljs-built_in">id</span> <span class="hljs-number">2142131</span> -<span class="hljs-built_in">format</span> txt<br></code></pre></div></td></tr></table></figure>

<p><strong>efetch</strong> 设置的参数是 <strong>-db</strong>，将搜索命中的记录从该数据库进行。如果 <strong>esearch</strong> 检索的结果和需要下载的结果来自同一数据库，可以不添加 <strong>-db</strong> 参数，系统会默认使用esearch的数据库。</p>
<p><strong>efetch</strong> 还要设置的参数是 <strong>-format</strong>，即下载数据的格式。常用的有 <strong>xml</strong> 格式，后续可以通过 <strong>xtract</strong> 和 <strong>linux</strong> 一些命令行工具来做处理。你也可以直接添加 <strong>-id</strong> 参数直接进行下载而不通过 <strong>esearch</strong> 检索后下载。如下面例子中将 <strong>pubmed id 2137412</strong> 的文献输出，格式为 Medline<a target="_blank" rel="noopener" href="https://indexofire.github.io/pathongs/book/C01_Common-Utility/03_entrez/#fn:2">3</a></p>
<figure class="highlight python"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs python"><span class="hljs-comment"># 搜索 pubmed 数据库编号 2137412 的文献，以 medline 格式输出</span><br>$ efetch -db pubmed -<span class="hljs-built_in">id</span> <span class="hljs-number">2137412</span> -<span class="hljs-built_in">format</span> medline<br></code></pre></div></td></tr></table></figure>

<p>对于 efetch 可以下载那些数据库，对于各个数据库名称和数据格式，可以用<code>-h</code>来查看</p>
<figure class="highlight elixir"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs elixir"><span class="hljs-variable">$ </span>efetch -h<br></code></pre></div></td></tr></table></figure>

<h4 id="2-3-xtract"><a href="#2-3-xtract" class="headerlink" title="2.3 xtract"></a>2.3 xtract</h4><p>edirect 的核心命令，由于 <strong>esearch</strong> 查询和下载的数据常常是 xml 格式的，对于 XML<a target="_blank" rel="noopener" href="https://indexofire.github.io/pathongs/book/C01_Common-Utility/03_entrez/#fn:3">4</a> 格式的数据，可以用 <strong>xtract</strong> 工具将其转换成纯文本格式的内容。xtract 不仅可以转换 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/">NCBI</a> 的 XML 格式文件，对于任何通用的 XML 格式文件都可以进行转换。</p>
<p>xtract 基本用法：</p>
<figure class="highlight pgsql"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs pgsql"># 读取 <span class="hljs-type">XML</span> 文件方式<br>$ xtract -<span class="hljs-keyword">input</span> <span class="hljs-type">XML</span>-file.xml<br></code></pre></div></td></tr></table></figure>

<p>一般我们都是通过管道工具 Pipeline 组合 esearch 和 efetch 等工具获得 XML 格式的 NCBI 数据库命中记录后，再用 xtract 进行处理。</p>
<figure class="highlight 1c"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs 1c">$ esearch -db pubmed -query &#x27;E. coli AND (<span class="hljs-number">2016/12/30</span>[pdat]:<span class="hljs-number">2016/12/31</span>[pdat])&#x27; <span class="hljs-string">| \</span><br>&gt; efetch -format xml <span class="hljs-string">| xtract -pattern PubmedArticle -element \</span><br>&gt; MedlineCitation/PMID Journal/ISOAbbreviation ArticleTitle<br></code></pre></div></td></tr></table></figure>

<p>这条命令将检索到的文献记录转换成行，再以 PMID，杂志名缩写和文章题目为列展开。命令打印的结果如下：</p>
<figure class="highlight mipsasm"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs mipsasm"><span class="hljs-number">28239441</span>J Public Health AfrQuality <span class="hljs-keyword">and </span>labeling information of Mo       ringa oleifera products marketed for HIV-infected people in Zimbabwe.<br><span class="hljs-number">28115890</span>Korean <span class="hljs-keyword">J </span>Food <span class="hljs-keyword">Sci </span>Anim ResourAntimicrobial Activity of Kefir against Variou     s Food Pathogens <span class="hljs-keyword">and </span>Spoilage <span class="hljs-keyword">Bacteria.</span><br><span class="hljs-keyword"></span><span class="hljs-number">28110668</span>Enzyme Microb. Technol  .A novel thermophilic <span class="hljs-keyword">and </span>halophilic esterase from <span class="hljs-keyword">Janibacter </span><span class="hljs-built_in">sp</span>. R02,   the first member of a new lipase family (Family XVII).<br>...<br><span class="hljs-number">27030192</span>J.   Vet. <span class="hljs-keyword">Sci.Development </span><span class="hljs-keyword">and </span>evaluation of an immunochromatographic assay   using a gp51 monoclonal antibody for the detection of antibodies against the <span class="hljs-keyword">bovine </span>leukemia virus.<br></code></pre></div></td></tr></table></figure>

<p>当我们不清楚XML元素结构时，可以用以下命令打印出树形结构。</p>
<figure class="highlight crmsh"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs crmsh">$ xtract -input my.<span class="hljs-keyword">xml</span> <span class="hljs-title">-outline</span><br></code></pre></div></td></tr></table></figure>

<p>xtract 可以灵活的改变数据的格式，所以涉及的参数也比较多。要想获得自己需要的数据格式，要想用好 edirect，掌握好 xtract 的用法是必不可少的。</p>
<p><strong>-pattern 参数</strong>: 将 xml 格式数据，将转换成行格式数据。每一个element包含的数据做为一行。</p>
<p>xtract 命令首先要通过 -pattern 定义所要分析的数据级别。比如对于 pubmed 数据库，efetch 输出 -format 格式为 docsum，那么我们一般要抓取的是每一个命中的文献记录，-pattern 就可以用 DocumentSummary 元素。如果你只想知道所有命中文献的作者（比如按照次数排序），那么就可以直接将 -pattern 定义为 Author。</p>
<p><strong>-element 参数</strong>: 将 <strong>-pattern</strong> 定义的行中以每个element形成一列。每个element以tab键分隔。</p>
<p>每个 -pattern 定义后，要多 -pattern 定义的元素的子元素进行输出，就要用到 -element 参数了。-element 可以跟多个子元素，默认以tab键分割。但是要注意的是子元素必须是唯一子元素，比如 -pattern 如果定义为 DocumentSummary，那么每个 DocumentSummary 元素有一个 Authors 子元素，而每个 Authors 有多个 Author 子元素（因为一篇文献有多个作者），如果直接将 -element Author，是无法返回作者信息的。需要用 -block 参数把 Authors 列转换成行再进行分割。</p>
<p><strong>-block 参数</strong>: 可以将某个XML标签下的多个相同标签转换成一行的多个元素，最常见的用法是在 <code>pubmed</code> 数据库检索文献时下多个标签打印出来。</p>
<figure class="highlight coq"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs coq">$ esearch -db pubmed -query <span class="hljs-string">&quot;Salmonella[ORGN]&quot;</span> -days <span class="hljs-number">5</span> | <span class="hljs-type">efetch</span> -format xml | <span class="hljs-type">\</span><br><span class="hljs-type">&gt; xtract</span> -<span class="hljs-built_in">pattern</span> PubmedArticle -block Author -sep <span class="hljs-string">&quot; &quot;</span> -element LastName,Initials<br></code></pre></div></td></tr></table></figure>

<p><strong>-tab、-sep 参数</strong>: 将多条 element 记录分割。</p>
<p><strong>/</strong>: 主要用在定义连续多层标签的定义上。比如PMID标签即可以在 MedlineCitation 标签下，有时也会在 RefSource 标签下。所以要明确检索文献的 PMID，就应该用 <strong>MedlineCitation/PMID</strong></p>
<p><strong>@</strong>: 指定 XML 元素 element 的 attribute 值</p>
<p><strong>-if、-if/-equals、-if/-contains 参数</strong>: 条件运算符</p>
<p><strong>-and、-or、-position、-def 参数</strong>: 关系运算符</p>
<h4 id="2-4-elink"><a href="#2-4-elink" class="headerlink" title="2.4 elink"></a>2.4 elink</h4><p>当我们要检索一个数据库数据所关联到其他数据库时，就要用到elink工具了。比如我们先在SRA数据库中搜索到所有鼠伤寒沙门菌的SRA实验数据，然后链接到biosample数据库，获得这些菌株的生物样本信息。</p>
<figure class="highlight stata"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs stata">$ esearch -<span class="hljs-keyword">db</span> sra -<span class="hljs-keyword">query</span> <span class="hljs-string">&quot;salmonella typhimurium&quot;</span> | elink -target biosample | efetch -<span class="hljs-keyword">format</span> docsum &gt; result.xml<br></code></pre></div></td></tr></table></figure>

<h4 id="2-4-einfo"><a href="#2-4-einfo" class="headerlink" title="2.4 einfo"></a>2.4 einfo</h4><p>对于 NCBI 数据库的各项复杂参数，字段等要完全了解和掌握比较困难。edirect 里提供了 einfo 工具，这样其他工具所能访问的具体数据库、字段和可链接数据库等信息都可以用 einfo 了解。</p>
<figure class="highlight powershell"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs powershell"><span class="hljs-comment"># 查询 `esearch -db` 参数所能访问的所有数据库名称</span><br><span class="hljs-variable">$</span> einfo <span class="hljs-literal">-dbs</span> | <span class="hljs-built_in">sort</span><br><br><span class="hljs-comment"># 查询某个数据库字段情况，如 sra 数据库所有对应的字段名称</span><br><span class="hljs-variable">$</span> einfo <span class="hljs-literal">-db</span> sra <span class="hljs-literal">-fields</span><br><br><span class="hljs-comment"># 查询数据库所有可链接的其他数据库</span><br><span class="hljs-variable">$</span> einfo <span class="hljs-literal">-db</span> sra <span class="hljs-literal">-links</span><br></code></pre></div></td></tr></table></figure>

<h3 id="3-高级检索"><a href="#3-高级检索" class="headerlink" title="3. 高级检索"></a>3. 高级检索</h3><p>工具 esearch 和 efilter 在做检索时如果做好筛选，能有助于获得准确的结果。筛选有2种方式，一种是通过添加参数来筛选。另一种是对 query 参数添加字段定义来筛选。</p>
<p><strong>参数过滤</strong>：</p>
<p>常用的参数为日期参数</p>
<ul>
<li>-days 100：100天内的检索条件命中</li>
<li>-maxdate：时间段最大日期，格式可以是 2001，2001/01/01</li>
<li>-mindate：时间段最小日期，格式可以是 2001，2001/01/01</li>
<li>-datetype: PDAT 发布时间；MDAT 修改时间</li>
</ul>
<p><strong>字段定义</strong>：</p>
<p>-query参数: esearch 和 efilter 通过添加该参数进行检索或过滤相应的关键词，query参数可以添加 filter，不同数据库可选择的filter可以用下面命令来查看。</p>
<figure class="highlight mipsasm"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs mipsasm"><span class="hljs-comment"># 查看sra数据库的filter</span><br>$ <span class="hljs-keyword">einfo </span>-db <span class="hljs-keyword">sra </span>-fields<br><br><span class="hljs-comment"># 查看sra数据库field，用xtract格式化输出</span><br>$ <span class="hljs-keyword">einfo </span>-db <span class="hljs-keyword">sra </span><span class="hljs-title">| xtract -pattern Field -element Name Description</span><br></code></pre></div></td></tr></table></figure>

<p>-query 检索词可以用逻辑关键词构建逻辑关系：</p>
<ul>
<li>AND：和</li>
<li>OR：或</li>
<li>NOT：非</li>
<li>()：共同定义</li>
</ul>
<p>-query 检索词还可以使用正则表达式：</p>
<ul>
<li>“*”星号表示任意个匹配的字符</li>
</ul>
<p><strong>-query 字符串定义</strong></p>
<p>建议用 ‘…’ 定义 -query 值，里面的各个关键词用 AND/OR 等建立逻辑关系。关键词组添加 filter 时用 “…” 定义。如：</p>
<p>-query ‘“salmonella paratyphi A”[ORGN] AND latest[filter] AND “complete genome[filter]”‘</p>
<h3 id="4-具体示例"><a href="#4-具体示例" class="headerlink" title="4. 具体示例"></a>4. 具体示例</h3><p>下面我们就用几个简单的例子来尝试 edirect 的功能。结合 Linux 其他命令行工具，可以极大的提升我们在命令行界面下检索和抓取 NCBI 数据库的效率。</p>
<p><strong>查看2014~2016年发表的霍乱弧菌CTX相关文献的摘要</strong></p>
<p>用关键词 vibrio cholerae CTX 检索 pubmed 数据库，将搜索到的文献的摘要保存到文件中，方便查看。由于是文本格式的内容，方便后期进一步处理，比如提取单词进行词频计算，用翻译API将摘要进行自动翻译等等。</p>
<figure class="highlight vala"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs vala"><span class="hljs-meta"># 示例为了减少命中数据，设置了发布日期的 -mindate 和 maxdata</span><br>$ esearch -db pubmed -mindate <span class="hljs-number">2014</span> -maxdate <span class="hljs-number">2016</span> -datetype PDAT \<br>&gt; -query <span class="hljs-string">&quot;vibrio cholerae CTX&quot;</span> | efetch -format <span class="hljs-keyword">abstract</span> &gt; <span class="hljs-keyword">abstract</span>.txt<br><br><span class="hljs-meta"># 日期设置的另一种写法</span><br>$ esearch -db pubmed -query <span class="hljs-string">&quot;vibrio cholerae CTX AND \</span><br><span class="hljs-string">&gt; (2014[pdat]:2016[pdat])&quot;</span> | efetch -format <span class="hljs-keyword">abstract</span> &gt; <span class="hljs-keyword">abstract</span>.txt<br></code></pre></div></td></tr></table></figure>

<p><strong>查看文章作者</strong></p>
<p>常常会对某一类文章，或者某一个期刊的文章作者进行关注，这里根据检索关键词，格式化输出文章的ID和作者列表。这也是 xtract 最常见的应用之一。</p>
<figure class="highlight node-repl"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs node-repl"># 检索关键词，以最近300天为范围<br>$ esearch -db pubmed -query &quot;listeria monocytogenes AND prfA&quot; -days 300 | \<br><span class="hljs-meta">&gt;</span> <span class="javascript">efetch -format xml | xtract -pattern PubmedArticle \</span><br><span class="hljs-meta">&gt;</span> <span class="javascript">-element MedlineCitation/PMID -block Author -sep <span class="hljs-string">&quot; &quot;</span> \</span><br><span class="hljs-meta">&gt;</span> <span class="javascript">-element LastName,Initials</span><br></code></pre></div></td></tr></table></figure>

<p><strong>查看近300天以来发表的文章中，作者中有叫Jim的所有文章题目</strong></p>
<p>利用 pubmed 数据库的 filter [FULL], 查询所有作者列表里有叫Jim的文章标题，这里采用 medline 格式输出，如果要进一步做格式化可以用xml格式再用xtract重新排版打印。</p>
<figure class="highlight stata"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs stata">$ esearch -<span class="hljs-keyword">db</span> pubmed -days 300 -<span class="hljs-keyword">query</span> <span class="hljs-string">&quot;Jim[FULL]&quot;</span> | efetch -<span class="hljs-keyword">format</span> medline | \<br>&gt; grep TI | awk -F&#x27;- &#x27; &#x27;&#123;<span class="hljs-keyword">print</span> <span class="hljs-variable">$2&#125;</span>&#x27;<br></code></pre></div></td></tr></table></figure>

<p><strong>查看2016年CDC发布的所有用miseq测序的沙门菌文库制备方法</strong></p>
<p>想要了解 CDC 开展的 miseq 沙门菌测序时采用的文库构建方法，可以利用 sra 数据库的 runinfo信息输出获得。</p>
<figure class="highlight awk"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs awk">$ esearch -db sra -query <span class="hljs-string">&quot;salmonella[OGRN] miseq[PLAT] EDBL-CDC&quot;</span> \<br>&gt; -mindate <span class="hljs-number">2016</span><span class="hljs-regexp">/01/</span><span class="hljs-number">01</span> -maxdate <span class="hljs-number">2016</span><span class="hljs-regexp">/12/</span><span class="hljs-number">31</span> -datetype PDAT | \<br>&gt; efetch -format runinfo | cut -d <span class="hljs-string">&#x27;,&#x27;</span> -f <span class="hljs-number">12</span> &gt; library.txt<br></code></pre></div></td></tr></table></figure>

<p><strong>绘制单增李斯特菌hlyA蛋白质进化树图</strong></p>
<figure class="highlight excel"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs excel">$ esearch -<span class="hljs-built_in">db</span> protein -query <span class="hljs-string">&quot;listeriolysin hlyA NOT partial&quot;</span> |<br>&gt; efetch -format fasta &gt; hlyA.fasta<br>$ mafft hlyA.fasta &gt; hlyA.mafft.fasta<br>$ raxml -m GTRGAMMA -p <span class="hljs-number">12345</span> -s hlyA.mafft.fasta -f a -x <span class="hljs-number">12345</span> -<span class="hljs-built_in">N</span> <span class="hljs-number">100</span> -<span class="hljs-built_in">T</span> <span class="hljs-number">40</span> -<span class="hljs-built_in">n</span> hlyA<br></code></pre></div></td></tr></table></figure>

<p><strong>对10年内发表的检索文献的作者排序</strong></p>
<figure class="highlight node-repl"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs node-repl">$ esearch -db pubmed -days 3650 -datetype PDAT -query &quot;hlyA&quot; | \<br><span class="hljs-meta">&gt;</span> <span class="javascript">efetch -format xml | \</span><br><span class="hljs-meta">&gt;</span> <span class="javascript">xtract -pattern Author -element LastName | \</span><br><span class="hljs-meta">&gt;</span> <span class="javascript">sort-uniq-count-rank</span><br></code></pre></div></td></tr></table></figure>

<p><strong>查看 taxonomy</strong></p>
<p>taxonomy 数据库的 filter [NXLV] 可以直接输出species级别</p>
<figure class="highlight stata"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs stata"># 获得 taxonomy 数据库里 <span class="hljs-keyword">lis</span> 或 sal 开头的分类名<br>$ esearch -<span class="hljs-keyword">db</span> taxonomy -<span class="hljs-keyword">query</span> <span class="hljs-string">&quot;lis* OR sal*&quot;</span> | efetch -<span class="hljs-keyword">db</span> taxonomy -<span class="hljs-keyword">format</span> txt<br><br># 获得 species 级的所有沙门菌<br>$ esearch -<span class="hljs-keyword">db</span> taxonomy -<span class="hljs-keyword">query</span> <span class="hljs-string">&quot;Salmonella[NXLV]&quot;</span> | efetch -<span class="hljs-keyword">db</span> taxonomy -<span class="hljs-keyword">format</span> txt<br></code></pre></div></td></tr></table></figure>

<p><strong>下载所有甲型副伤寒沙门菌基因组序列</strong></p>
<figure class="highlight stata"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs stata">$ esearch -<span class="hljs-keyword">db</span> nuccore -<span class="hljs-keyword">query</span> <span class="hljs-string">&quot;Salmonella[porgn:__txid54388] complete genome&quot;</span> | \<br>&gt; efetch -<span class="hljs-keyword">db</span> nuccore -<span class="hljs-keyword">format</span> fasta &gt; SPA.fasta<br></code></pre></div></td></tr></table></figure>

<p><strong>获取SRA数据库中目的数据的测序样本地理信息</strong></p>
<figure class="highlight node-repl"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs node-repl">$ esearch -db sra -query &quot;China&quot; -days 1 | elink -target biosample | \<br><span class="hljs-meta">&gt;</span> <span class="javascript">efetch -format docsum | xtract -pattern DocumentSummary -element Accession \</span><br><span class="hljs-meta">&gt;</span> <span class="javascript">-block Attribute -<span class="hljs-keyword">if</span> Attribute@attribute_nam \</span><br><span class="hljs-meta">&gt;</span> <span class="javascript">-equals lat_lon -element Attribute</span><br></code></pre></div></td></tr></table></figure>

<p><strong>获得某个bioproject下的所有sra run的信息</strong></p>
<figure class="highlight coq"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs coq">$ eserach -db bioproject -query <span class="hljs-string">&quot;PRJNA209644&quot;</span> | <span class="hljs-type">elink</span> -target sra | <span class="hljs-type">\</span><br><span class="hljs-type">&gt; efetch</span> -format docsum | <span class="hljs-type">xtract</span> -<span class="hljs-built_in">pattern</span> DocumentSummary -ACC @acc \<br>&gt; -block DocumentSummary -element <span class="hljs-string">&quot;&amp;ACC&quot;</span><br></code></pre></div></td></tr></table></figure>

<p><strong>下载assembly数据库的refseq序列</strong></p>
<figure class="highlight stata"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs stata">$ esearch -<span class="hljs-keyword">db</span> assembly -<span class="hljs-keyword">query</span> <span class="hljs-string">&quot;Salmonella enterica[ORGN] AND 2010[GRLS]&quot;</span> | \<br>&gt; efetch -<span class="hljs-keyword">format</span> docsum | xtract -pattern DocumentSummary \<br>&gt; -element FtpPath_RefSeq | awk -F<span class="hljs-string">&quot;/&quot;</span> &#x27;&#123;<span class="hljs-keyword">print</span> <span class="hljs-variable">$0</span><span class="hljs-string">&quot;/&quot;</span><span class="hljs-variable">$NF</span><span class="hljs-string">&quot;_genomic.fna.gz&quot;</span>&#125;&#x27; | \<br>&gt; xargs wget<br></code></pre></div></td></tr></table></figure>

<p><strong>查看assembly meta信息</strong></p>
<p>NCBI 的 Assembly 数据库包含了 RefSeq 数据库的基因组数据，选择相应的物种可以获得该物种的基因组拼接序列，有完成图也有非完成图。</p>
<figure class="highlight livescript"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs livescript">$ esearch -db assembly -query <span class="hljs-string">&#x27;&quot;salmonella paratyphi A&quot;[ORGN] AND latest[FILT]&#x27;</span> | <span class="hljs-string">\</span><br>&gt; efetch -format docsum | xtract -pattern Meta -block Stats <span class="hljs-string">\</span><br>&gt; -tab <span class="hljs-string">&#x27;\n&#x27;</span> -sep <span class="hljs-string">&#x27;\t&#x27;</span> -element Stat@category Stat | sort -ur &gt; meta.txt<br><span class="hljs-keyword">import</span> pandas <span class="hljs-keyword">as</span> pd<br><span class="hljs-keyword">import</span> seaborn <span class="hljs-keyword">as</span> sns<br><br>df = pd.read_table(<span class="hljs-string">&quot;meta.txt&quot;</span>)<br>sns.jointplot(x=<span class="hljs-string">&quot;contig_count&quot;</span>, y=<span class="hljs-string">&quot;scaffold_n50&quot;</span>, data=df)<br></code></pre></div></td></tr></table></figure>

<p><strong>获取SRA菌株信息</strong></p>
<p>对某个物种开展进化研究时，我们往往会竟可能收集公共数据库中的菌株背景信息。下面这个例子搜索某个 bioproject，然后获得该研究的所有测序数据的 sra 背景信息。通过 &lt;SAMPLE_ATTRIBUTE&gt; 元素下的各个 TAG 和 VALUE 来提取所需信息。</p>
<figure class="highlight stata"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs stata">$ esearch -<span class="hljs-keyword">db</span> bioproject -<span class="hljs-keyword">query</span> 273159 | elink -target sra | \<br>&gt; efetch -<span class="hljs-keyword">format</span> xml | xtract -pattern EXPERIMENT_PACKAGE -<span class="hljs-keyword">tab</span> <span class="hljs-string">&quot;,&quot;</span> \<br>&gt; -element <span class="hljs-keyword">RUN</span>@accession -block SAMPLE_ATTRIBUTE -<span class="hljs-keyword">tab</span> <span class="hljs-string">&quot;\t&quot;</span> -sep <span class="hljs-string">&quot;:&quot;</span> \<br>&gt; -element TAG,VALUE &gt; <span class="hljs-keyword">sample</span>.txt<br>$ head <span class="hljs-keyword">sample</span>.txt<br></code></pre></div></td></tr></table></figure>

<p>这样输出的问题在于各个测序数据的 SAMPLE_ATTRIBUTE 的 TAG 顺序是乱序的，每个 SRR 输出的顺序不同导致各个字段要重新排序，比较麻烦。如果我们只需要其中一个字段，可以用 -if *** -eqauls *** 来过滤。比如下面代码我们输出菌株名称：</p>
<figure class="highlight crmsh"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs crmsh">$ esearch -db bioproject -query <span class="hljs-number">273159</span> | elink -target sra | \<br>&gt; efetch -format <span class="hljs-keyword">xml</span> <span class="hljs-title">| xtract</span> -pattern EXPERIMENT_PACKAGE -tab <span class="hljs-string">&quot;,&quot;</span> \<br>&gt; -element RUN@accession -block SAMPLE_ATTRIBUTE -if <span class="hljs-keyword">TAG</span> <span class="hljs-title">-equals</span> <span class="hljs-string">&quot;strain&quot;</span> \<br>&gt; -element VALUE | sort &gt; strain-name.csv<br></code></pre></div></td></tr></table></figure>

<p><code>-if</code>在一行输出中不能反复使用，如果要同时对多个样品信息字段分别输出，我们采用变通的方法，先把 xml 数据保存到本地，然后循环分别输出各个所需内容，用 sort 排序，再用 paste 合并。</p>
<figure class="highlight crmsh"><table><tr><td class="gutter hljs"><div class="hljs code-wrapper"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></div></td><td class="code"><div class="hljs code-wrapper"><pre><code class="hljs crmsh"><span class="hljs-comment"># 分别输出 csv 格式的数据</span><br>$ for i <span class="hljs-keyword">in</span> serovar strain isolate_name_alias collection_date isolation_source; \<br>&gt; do xtract -input bioproject.<span class="hljs-keyword">xml</span> <span class="hljs-title">-pattern</span> EXPERIMENT_PACKAGE -tab <span class="hljs-string">&quot;,&quot;</span> \<br>&gt; -element RUN@accession -block SAMPLE_ATTRIBUTE -if <span class="hljs-keyword">TAG</span> <span class="hljs-title">-equals</span> $i -element \<br>&gt; VALUE | sort &gt; $i; done<br><span class="hljs-comment"># 合并 csv</span><br>$ paste -d , strain serovar isolate_name_alias collection_date &gt; sample.csv<br></code></pre></div></td></tr></table></figure>

<h3 id="Reference"><a href="#Reference" class="headerlink" title="Reference"></a>Reference</h3><hr>
<ol>
<li><a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/books/NBK179288">Entrez Direct: E-utilities on the UNIX Command Line</a> <a target="_blank" rel="noopener" href="https://indexofire.github.io/pathongs/book/C01_Common-Utility/03_entrez/#fnref:1">↩</a></li>
<li><a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Pipeline_(Unix)">Unix Pipeline Wikipedia</a> <a target="_blank" rel="noopener" href="https://indexofire.github.io/pathongs/book/C01_Common-Utility/03_entrez/#fnref:4">↩</a></li>
<li><a target="_blank" rel="noopener" href="https://www.nlm.nih.gov/bsd/licensee/elements_descriptions.html">Medline Element Description</a> <a target="_blank" rel="noopener" href="https://indexofire.github.io/pathongs/book/C01_Common-Utility/03_entrez/#fnref:2">↩</a></li>
<li><a target="_blank" rel="noopener" href="http://www.w3school.com.cn/xml/">XML</a> <a target="_blank" rel="noopener" href="https://indexofire.github.io/pathongs/book/C01_Common-Utility/03_entrez/#fnref:3">↩</a></li>
</ol>

            </div>
            <hr>
            <div>
              <div class="post-metas mb-3">
                
                
              </div>
              
              
                <div class="post-prevnext">
                  <article class="post-prev col-6">
                    
                    
                      <a href="/2021/09/12/seqkit/">
                        <i class="iconfont icon-arrowleft"></i>
                        <span class="hidden-mobile">seqkit</span>
                        <span class="visible-mobile">上一篇</span>
                      </a>
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/2021/09/12/xiazai/">
                        <span class="hidden-mobile">下载基因组数据</span>
                        <span class="visible-mobile">下一篇</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
          </article>
        </div>
      </div>
    </div>
    
      <div class="d-none d-lg-block col-lg-2 toc-container" id="toc-ctn">
        <div id="toc">
  <p class="toc-header"><i class="iconfont icon-list"></i>&nbsp;目录</p>
  <div class="toc-body" id="toc-body"></div>
</div>

      </div>
    
  </div>
</div>

<!-- Custom -->


    

    
      <a id="scroll-top-button" aria-label="TOP" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">搜索</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v"
                 for="local-search-input">关键词</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>
    

    
  </main>

  <footer class="text-center mt-5 py-3">
  <div class="footer-content">
     <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>科研民工</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>林国鹏</span></a> 
  </div>
  

  

  
</footer>


  <!-- SCRIPTS -->
  
  <script  src="https://cdn.jsdelivr.net/npm/nprogress@0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/nprogress@0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://cdn.jsdelivr.net/npm/jquery@3/dist/jquery.min.js" ></script>
<script  src="https://cdn.jsdelivr.net/npm/bootstrap@4/dist/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>

<!-- Plugins -->


  <script  src="/js/local-search.js" ></script>



  
    <script  src="/js/img-lazyload.js" ></script>
  



  



  
    <script  src="https://cdn.jsdelivr.net/npm/tocbot@4/dist/tocbot.min.js" ></script>
  
  
    <script  src="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3/dist/jquery.fancybox.min.js" ></script>
  
  
    <script  src="https://cdn.jsdelivr.net/npm/anchor-js@4/anchor.min.js" ></script>
  
  
    <script defer src="https://cdn.jsdelivr.net/npm/clipboard@2/dist/clipboard.min.js" ></script>
  






  <script  src="https://cdn.jsdelivr.net/npm/typed.js@2/lib/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var title = document.getElementById('subtitle').title;
      
      typing(title)
      
    })(window, document);
  </script>















<!-- 主题的启动项 保持在最底部 -->
<script  src="/js/boot.js" ></script>


</body>
</html>
